gh-77024: test.support: Improve documentation (#92513)
python/cpython: Doc/library/test.rst

   to make writes blocking.

.. data:: TEST_SUPPORT_DIR

>    Set to the top level directory that contains :mod:`test.support`.

Reviewer: Why do you remove this one? It still exists, no? If you want to remove the variable itself, I suggest to open a separated issue. TEST_SUPPORT_DIR is not used anywhere in Python.

Author: It's used only to compute TEST_HOME_DIR, so I figured it wasn't worth documenting. I can put it back if you prefer.

.. data:: MISSING_C_DOCSTRINGS

   Set to ``True`` if running on CPython, not on Windows, and configuration

>    not set with ``WITH_DOC_STRINGS``.

Reviewer: I suggest:

```
Set to ``True`` if Python is built without docstrings: if the :c:macro:`WITH_DOC_STRINGS` macro is not defined.
See the :option:`configure --without-doc-strings <--without-doc-strings>` option.

See also the :data:`HAVE_DOCSTRINGS` variable.
```

I don't think that it's useful to mention Windows here.

Author: Thanks for the review!

.. data:: HAVE_DOCSTRINGS

>    Set to ``True`` if docstrings are enabled.

Reviewer: ```suggestion
   Set to ``True`` if function docstrings are available.
   See the :option:`python -OO <-O>` which strips docstrings of functions implemented in Python.
   
   See also the :data:`MISSING_C_DOCSTRINGS` variable.
```

.. data:: MISSING_C_DOCSTRINGS

   Set to ``True`` if Python is built without docstrings (the
   :c:macro:`WITH_DOC_STRINGS` macro is not defined).
   See the :option:`configure --without-doc-strings <--without-doc-strings>` option.

   See also the :data:`HAVE_DOCSTRINGS` variable.

.. data:: HAVE_DOCSTRINGS

   Set to ``True`` if function docstrings are available.

>    See the :option:`python -OO <-O>` which strips docstrings of functions implemented in Python.

Reviewer 2: ```suggestion
   See the :option:`python -OO <-O>` option, which strips docstrings of functions implemented in Python.
```

.. function:: match_test(test)

   Return whether *test* matches the patterns set in :func:`set_match_tests`.

.. function:: set_match_tests(accept_patterns=None, ignore_patterns=None)

>    Define match patterns for filtering tests.

Reviewer: ```suggestion
   Define match patterns on test filenames and test method names for filtering tests.
```

.. function:: match_test(test)

>    Return whether *test* matches the patterns set in :func:`set_match_tests`.

Reviewer 2: ```suggestion
   Determine whether *test* matches the patterns set in :func:`set_match_tests`.
```

.. function:: set_match_tests(accept_patterns=None, ignore_patterns=None)

>    Define match patterns on test filenames and test method names for filtering tests.

Reviewer 2: I think this one is a little too terse; something like this? (I've never used this function.)

```suggestion
   Register :mod:`regular expression <re>` patterns for filtering tests. Tests are filtered based on whether filenames or test method names match one or more of the patterns.
```

Author: That's still incomplete, as it doesn't explain how accept_patterns and ignore_patterns interact. I haven't used the function either, so I'd prefer to stick with the version Victor signed off on.

Reviewer 2: Sure, let's leave this for now then

.. function:: disable_gc()

   A context manager that disables the garbage collector upon entry and

>    reenables it upon exit, if it was initially enabled.

Reviewer 2: (GitHub's not letting me do a suggested change because the line range spans a deleted line.)

I would say something like:

```
   A context manager that disables the garbage collector upon entry. On
   exit, the garbage collector is restored to its prior state.
```

Author: Thanks, applying

.. function:: calcobjsize(fmt)

>    Return the size of the :c:type:`PyObject` whose structure is

Reviewer: ```suggestion
   Return the size of the :c:type:`PyObject` whose structure members are
```

Same remark for calcvobjsize().

.. function:: calcobjsize(fmt)

   Return the size of the :c:type:`PyObject` whose structure members are

>    defined by *fmt* including the Python object header and alignment.

Reviewer 2: I would split this (and `calcvobjsize` below) into two sentences, personally.

```suggestion
   defined by *fmt*. The returned value includes the size of the Python object header and alignment.
```

.. function:: calcvobjsize(fmt)

   Return the size of the :c:type:`PyVarObject` whose structure members are

>    defined by *fmt* including the Python object header and alignment.

Author: ```suggestion
   defined by *fmt*. The returned value includes the size of the Python object header and alignment.
```

   have an associated comment identifying the relevant tracker issue.

.. function:: system_must_validate_cert(f)

>    A decorator that skips the test on TLS certification validation failures.

Reviewer 2: ```suggestion
   A decorator that skips the decorated test on TLS certification validation failures.
```

   variables *env_vars* succeeds (``rc == 0``) and return a ``(return code,
   stdout, stderr)`` tuple.

>    If the *__cleanenv* keyword is set, *env_vars* is used as a fresh

Reviewer 2: ```suggestion
   If the *__cleanenv* keyword-only parameter is set, *env_vars* is used as a fresh
```

   environment.

   Python is started in isolated mode (command line option ``-I``),

>    except if the *__isolated* keyword is set to ``False``.

Reviewer 2: ```suggestion
   except if the *__isolated* keyword-only parameter is set to ``False``.
```

   is still alive after *timeout* seconds.

.. decorator:: reap_threads

   Decorator to ensure the threads are cleaned up even if the test fails.

.. function:: start_threads(threads, unlock=None)

   Context manager to start *threads*, which is a sequence of threads.
   *unlock* is a function called after the threads are started, even if an
   exception was raised; an example would be :meth:`threading.Event.set`.

>    This also attempts to join the started threads upon exit.

Reviewer 2: Unclear whether "this" refers to ``start_threads`` or ``threading.Event.set```

```suggestion
   ``start_threads`` will attempt to join the started threads upon exit.
```

.. data:: TESTFN_NONASCII

   Set to a filename containing the :data:`FS_NONASCII` character, if it exists.
   This guarantees that if the filename exists, it can be encoded and decoded
   with the default filesystem encoding. This allows tests that require a

>    non-ASCII filename to be easily skipped on platforms where they can't work.

Reviewer 2: This currently feels like it leads with the implementation, and finishes with why it's useful ("This allows tests that require a non-ASCII filename to be easily skipped on platforms where they can't work"). The other way around might be better.

Author: I see your point, but I feel like the current sequence is more in line with the surrounding TESTFN_* paragraphs.

   Call :func:`os.rmdir` on *filename*.  On Windows platforms, this is

>    wrapped with a wait loop that checks for the existence of the file.
>    This is needed due to antivirus programs that can hold files open and prevent

Reviewer 2: ```suggestion
   wrapped with a wait loop that checks for the existence of the file,
   which is needed due to antivirus programs that can hold files open and prevent
```
